java/core/src/main/java/b3/TraceContext.java

```diff
@@ -0,0 +1,357 @@
+    if (bytes.length != 32 && bytes.length != 40) {
+      throw new IllegalArgumentException("bytes.length " + bytes.length + " != 32 or 40");
+    }
+
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    long spanId = buffer.getLong(0);
+    long parentId = buffer.getLong(8);
+    TraceId traceId;
+    long flags;
+    if (bytes.length == 32) {
+      long lo = buffer.getLong(16);
+      traceId = new TraceId(0, (parentId == lo) ? parentId : lo);
+      flags = buffer.getLong(24);
+    } else {
+      traceId = new TraceId(
+          buffer.getLong(16),
+          buffer.getLong(24)
+      );
+      flags = buffer.getLong(32);
+    }
+
+    return new TraceContext(traceId, parentId, spanId, flags);
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Unique 8-byte identifier for a trace, set on all spans within it.
+   */
+  public final TraceId traceId;
+
+  /**
+   * The parent's {@link #spanId} or {@link #spanId} if this the root span in a trace.
+   */
+  public final long parentId;
+
+  /** Returns null when this is a root span. */
+  public Long nullableParentId() {
+    return root() ? null : parentId;
+  }
+
+  /**
+   * Unique 8-byte identifier of this span within a trace.
+   *
+   * <p>A span is uniquely identified in storage by ({@linkplain #traceId}, {@code #id}).
+   */
+  public final long spanId;
+
+  /** Returns true if this is the root span. */
+  public final boolean root() {
+    return (flags & FLAG_IS_ROOT) == FLAG_IS_ROOT || parentId == traceId.lo && parentId == spanId;
+  }
+
+  /**
+   * True is a request to store this span even if it overrides sampling policy. Implies {@link
+   * #nullableSampled()}.
+   */
+  public final boolean debug() {
+    return (flags & FLAG_DEBUG) == FLAG_DEBUG;
+  }
+
+  /**
+   * Should we sample this request or not? True means sample, false means don't, null means we defer
+   * decision to someone further down in the stack.
+   */
+  public final Boolean nullableSampled() {
+    if (debug()) return true;
+    return (flags & FLAG_SAMPLING_SET) == FLAG_SAMPLING_SET
+        ? (flags & FLAG_SAMPLED) == FLAG_SAMPLED
+        : null;
+  }
+
+  /** Raw flags encoded in {@link #bytes()} */
+  public final long flags;
+
+  /** Serializes this into a big-endian byte array */
+  public byte[] bytes() {
+    boolean traceHi = traceId.hi != 0;
+    byte[] result = new byte[traceHi ? 40 : 32];
+    ByteBuffer buffer = ByteBuffer.wrap(result);
+    buffer.putLong(0, spanId);
+    buffer.putLong(8, parentId);
+    if (traceHi) {
+      buffer.putLong(16, traceId.hi);
+      buffer.putLong(24, traceId.lo);
+      buffer.putLong(32, flags);
+    } else {
+      buffer.putLong(16, traceId.lo);
+      buffer.putLong(24, flags);
+    }
+    return result;
+  }
+
+  public Builder toBuilder() {
+    return new Builder(this);
+  }
+
+  /** Returns {@code $traceId.$spanId<:$parentId} */
+  @Override
+  public String toString() {
+    boolean traceHi = traceId.hi != 0;
```

Comment:
here and elsewhere I treat traceId.hi == 0 as backwards compat mode.. exactly the same logic and encoding as before.